### ISSUE TITLE
Backprop with overwriting mode (no accumulation)

### ISSUE THREAD

**Issue by beam2d** [chainer/chainer#2055]
We often do not need to accumulate the gradient to parameters. Currently, we have to clear the gradient arrays before backprop, but if we have an interface to force overwriting the gradients, using it would be an easier solution.

**Comment by delta2323** [comment]
I would like to ask you why this feature is considered as `v2`. I thought `v2` features are ones that would breaks compatibility, but I have not come up with compatibility that would be broken by this feature so far, if the default behavior is to accumulate gradients.

**Comment by beam2d** [comment]
It's ok to implement it for v1. I just meant that the design of this feature would need a time to review and thus v2 would be a good deadline, but if that's not the case and it is a task to just implement, there is no concern about make it available in a minor update.

**Comment by delta2323** [comment]
OK, I understand your intention. I will design to see the rough estimate of the time required. For now, I think v2 is a good deadline, too.

**Comment by delta2323** [comment]
Do you have some examples/use cases in which need to clean up gradients before back propagation? It would be helpful.

**Comment by beam2d** [comment]
For example, GradientMethod calls cleargrads or zerograds before backprop. When users write their own optimization routine manually, they also have to call either before backprop to avoid accumulating the new gradient to the old one.

**Comment by delta2323** [comment]
Do you mean overloading `update` method by "write own optimization routine"?

**Comment by beam2d** [comment]
That's one example, but it is not limited to that case. For example, users can call backward manually and then call `GradientMethod.update()` by passing nothing. In this case, whether the gradient is correctly reset is up to users.

**Comment by delta2323** [comment]
Thank you for answering my question. By the way, do you know why most of DL frameworks (as far as I know) take the strategy of accumulating gradients, rather than overwriting it by default although in the typical procedure, back propagation is done only once for each mini batch?

**Comment by delta2323** [comment]
I would like to clear where the problem is in the current implementation. 

You mentioned in the initial issue comment that the motivation for the gradient overwriting mode is to alleviate the cumbersome of users to clear gradients by themselves. I am wondering if the introduction of overwriting mode can solve the problem. Because it depends on use cases whether the correct behavior is to accumulate gradients or to overwrite gradients, users have to choose which should be considered as correct anyway. I think calling `cleargrads` or `zerograds` before `backprop` is almost easiest way to do so.

A possible improvement is to make "overwriting mode" and "accumulating mode" symmetric e.g by making flags that switches these two modes or add option to `Variable.backward` that specifies the mode.

I am skeptical if I can fully understand the motivation for this PR. Let me know if I misunderstand something.

**Comment by beam2d** [comment]
> do you know why most of DL frameworks (as far as I know) take the strategy of accumulating gradients, rather than overwriting them by default although in the typical procedure, back propagation is done only once for each mini batch?

I don't think most of DL frameworks accumulate gradients. Most frameworks based on static graph does not accumulate gradients to the existing one (since the gradients are separate nodes in their graphs). For dynamic graph frameworks, accumulating gradients makes the framework implementation simpler, since the gradient must be accumulated in any branching points of the computational graph and thus we eventually needs to implement accumulation in some way.

**Comment by delta2323** [comment]
I had Caffe and Torch.nn in my mind, which do not create computational graphs for backwards. I thought Caffe accumulates gradients when reading e.g. [here](https://github.com/BVLC/caffe/blob/master/src/caffe/layers/inner_product_layer.cpp#L108) and Torch also accumulate it from the existence of [zeroGradParameters](https://github.com/torch/nn/blob/master/doc/module.md#nn.Module.zeroGradParameters). I would like to be pointed out if my reasoning is wrong. Anyway, it was exaggerate to say "most of DL frameworks".

As for the static-graph-based frameworks which construct graphs for backward and dynamic-graph-based frameworks, your explanation is reasonable to me. Thank you for the explanation.

**Comment by delta2323** [comment]
I remove the v2 label as this issue need not break compatibility and is relatively low priority.
